server/src/main/java/org/elasticsearch/search/vectors/ScoreDocQueryBuilder.java

```diff
@@ -0,0 +1,156 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+package org.elasticsearch.search.vectors;
+
+import org.apache.lucene.index.IndexReader;
+import org.apache.lucene.search.Query;
+import org.apache.lucene.search.ScoreDoc;
+import org.elasticsearch.Version;
+import org.elasticsearch.common.io.stream.StreamInput;
+import org.elasticsearch.common.io.stream.StreamOutput;
+import org.elasticsearch.common.lucene.Lucene;
+import org.elasticsearch.index.query.AbstractQueryBuilder;
+import org.elasticsearch.index.query.SearchExecutionContext;
+import org.elasticsearch.xcontent.XContentBuilder;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Comparator;
+import java.util.List;
+import java.util.Objects;
+
+/**
+ * A query that matches the provided docs with their scores. This query is used
+ * when executing a kNN search during the search query phase, to include the documents
+ * that matched the initial kNN query during the DFS phase.
+ */
+public class ScoreDocQueryBuilder extends AbstractQueryBuilder<ScoreDocQueryBuilder> {
+    public static final String NAME = "score_doc";
```

Comment:
Do you think we should give more descriptive name (e.g. by adding knn) to this query in case it is used for debugging (e.g. slow log).?

Reply:
👍 maybe we could do `knn_score_doc`. I think it's unlikely this query will be useful in a different context.